sklearn/tree/_criterion.pyx

```diff
@@ -962,6 +963,162 @@ cdef class MSE(RegressionCriterion):
         impurity_left[0] /= self.n_outputs
         impurity_right[0] /= self.n_outputs
 
+cdef class MAE(RegressionCriterion):
+    """Mean absolute error impurity criterion
+    """
+    cdef void node_value(self, double* dest) nogil:
+        """Computes the node value of samples[start:end] into dest."""        
+        cdef SIZE_t start = self.start
+        cdef SIZE_t end = self.end
+
+        cdef double* y_vals = <double*> calloc(self.n_node_samples, sizeof(double))
+        cdef double* weights = <double*> calloc(self.n_node_samples, sizeof(double))
+        self.compute_weighted_median(dest, y_vals, weights, start, end)
+
+    cdef void compute_weighted_median(self, double* median_dest, double* y_vals,
+                                      double* weights, SIZE_t start, SIZE_t end) nogil:
+        """Calculate the weighted median and put it into a destination pointer
+        given values, weights, and a start and end index
+        """
+        cdef double* sample_weight = self.sample_weight
+        cdef DOUBLE_t* y = self.y
+        cdef SIZE_t* samples = self.samples
+        cdef DOUBLE_t w = 1.0
+        cdef DOUBLE_t y_ik
+
+        cdef SIZE_t i, p, k
+        cdef DOUBLE_t sum_weights
+        cdef SIZE_t median_index
+        cdef DOUBLE_t sum
+
+        for k in range(self.n_outputs):
+            median_index = 0
+            sum_weights = 0.0
+            for p in range(start,end):
+                i = samples[p]
+                
+                y_ik = y[i * self.y_stride + k]
+                if sample_weight != NULL:
+                    w = sample_weight[i]
+
+                y_vals[p] = y_ik
+                weights[p] = w
+        
+            for p in range(start, end):
+                sum_weights += weights[p]
+
+            self.sort_values_and_weights(y_vals, weights, 0, self.n_node_samples - 1)
+            
+            sum = sum_weights - weights[0]
+            
+            while(sum > sum_weights/2):
+                median_index +=1
+                sum -= weights[median_index]
+
+            if sum == sum_weights/2:
+                median_dest[k] = (y_vals[median_index] + y_vals[median_index + 1]) / 2
+            else:
+                median_dest[k] = y_vals[median_index]
+            
+                
+    cdef void sort_values_and_weights(self, double* y_vals, double* weights,
```

Comment:
I've changed the double / DOUBLE_t as appropriate, can you let me know if I missed anything?